src/action/base/create_or_merge_nix_config.rs

```diff
@@ -0,0 +1,505 @@
+        let span = span!(
+            tracing::Level::DEBUG,
+            "create_file",
+            path = tracing::field::display(self.path.display()),
+            user = self.user,
+            group = self.group,
+            mode = self
+                .mode
+                .map(|v| tracing::field::display(format!("{:#o}", v))),
+            buf = tracing::field::Empty,
+        );
+
+        if tracing::enabled!(tracing::Level::TRACE) {
+            span.record("buf", &self.buf);
+        }
+        span
+    }
+
+    fn execute_description(&self) -> Vec<ActionDescription> {
+        vec![ActionDescription::new(self.tracing_synopsis(), vec![])]
+    }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn execute(&mut self) -> Result<(), ActionError> {
+        let Self {
+            path,
+            user,
+            group,
+            mode,
+            buf,
+            nix_configs,
+        } = self;
+
+        if tracing::enabled!(tracing::Level::TRACE) {
+            let span = tracing::Span::current();
+            span.record("buf", &buf);
+        }
+
+        let mut options = OpenOptions::new();
+        options.create(true).write(true).read(true);
+
+        if let Some(mode) = mode {
+            options.mode(*mode);
+        }
+
+        let mut file = options
+            .open(&path)
+            .await
+            .map_err(|e| ActionError::Open(path.to_owned(), e))?;
+
+        if let Some(merged_nix_config) = &nix_configs.merged_nix_config {
+            // FIXME(@cole-h): for now we replace the entire file, but in the future we could potentially "replace" the contents
+            let mut new_config = String::new();
+            if let Some(existing_nix_config) = &nix_configs.existing_nix_config {
+                for (name, value) in existing_nix_config {
+                    if merged_nix_config.get(name).is_some() {
+                        continue;
+                    }
+
+                    new_config.push_str(name);
+                    new_config.push_str(" = ");
+                    new_config.push_str(value);
+                    new_config.push('\n');
+                }
+
+                new_config.push('\n');
+            }
+
+            new_config.push_str(&format!(
+                    "# Generated by https://github.com/DeterminateSystems/nix-installer, version {version}.\n",
+                    version = env!("CARGO_PKG_VERSION"),
+                ));
+
+            for (name, value) in merged_nix_config {
+                new_config.push_str(name);
+                new_config.push_str(" = ");
+                new_config.push_str(value);
+                new_config.push('\n');
+            }
+
+            file.rewind()
+                .await
+                .map_err(|e| ActionError::Seek(path.to_owned(), e))?;
+            file.set_len(0)
+                .await
+                .map_err(|e| ActionError::Truncate(path.to_owned(), e))?;
```

Comment:
I wonder if this would be a good place to do similar to @lheckemann 's work in on `CreateOrInsertIntoFile` which does an atomic move

Reply:
I was actually thinking the same thing. I'll add it to my TODO list!